include/shark/Core/Threading/Pipeline.h

```diff
@@ -0,0 +1,84 @@
+#ifndef SHARK_CORE_THREADING_PIPELINE_H
+#define SHARK_CORE_THREADING_PIPELINE_H
+
+#include <shark/Core/Threading/ThreadPool.h>
+#include <shark/Core/utility/KeyValuePair.h>
+#include <queue>
+#include <mutex>
+namespace shark{
+namespace threading{
+
+/// \brief a Pipeline computes an ordered set of tasks using a threadpool.
+///
+/// For a set of tasks T1...TN computes Xi=f(Ti). the computations are assumed to be independent
+/// and are performed in parallel. However, the results Xi are kept in the correct order. In this sense,
+/// the pipeline is a true FIFO-Queue.
+/// Note that pushing and pulling elements in the pipeline is not threadsafe.
+template<class T>
+class Pipeline{
+public:
+	typedef T value_type;
+	Pipeline(std::size_t maxUnreadElements, ThreadPool& threadpool)
+	: m_threadpool(threadpool)
+	, m_maxUnreadElements(maxUnreadElements)
+	, m_task_start_counter(0)
+	, m_task_read_counter(0){}
+	
+	bool empty(){
+		return m_task_start_counter == m_task_read_counter;
+	}
+	
+	template<class Task>
+	queue_status push(Task&& task){
```

Comment:
This call is not moving the argument. More like indicating that it might be moved (so that it can bind with temporary values). I should add as documentation, that push/pull will not change contents of the supplied argument, unless the call was successful.